File: Algorithms/Stack/394. Decode String.py
#regular expression
import re
class Solution(object):
	def decodeString(self, s):
		pattern = re.compile(r'(\d+)\[([a-z]*)\]')	
		while('[' in s):
			s = pattern.sub(lambda x: int(x.group(1)) * x.group(2), s)
		return s

#stack
class Solution(object):
	def decodeString(self, s):
		stack = [['', 1]]
		num = ''
		for c in s:
			if(c.isdigit()):
				num += c
			elif(c == '['):
				stack.append(['', int(num)])	
				num = ''
			elif(c == ']'):
				string, count = stack.pop()
				stack[-1][0] += count * string
			else:
				stack[-1][0] += c
		return stack[-1][0]

File: Algorithms/String/394. Decode String.java
class Solution {
    public String decodeString(String s) {
        Stack<String> stack = new Stack();
        String cur = "", num = "";
        
        for(int i = 0; i < s.length(); ++i) {
            char c = s.charAt(i);
            if(Character.isDigit(c)) {
                num += c;
            }
            else if(c == '[') {
                stack.push(cur);
                stack.push(num);
                num = "";
                cur = "";
            }
            else if(c == ']') {
                int count = Integer.parseInt(stack.pop());
                String pre = stack.pop();
                cur = cur.repeat(count);
                cur = pre + cur;
            }
            else {
                cur += c;
            }
        }
        return cur;
    }
}
